app/models/manageiq/providers/vmware/inventory/parser/cloud_manager.rb

```diff
@@ -0,0 +1,119 @@
+class ManageIQ::Providers::Vmware::Inventory::Parser::CloudManager < ManageIQ::Providers::Vmware::Inventory::Parser
+  def parse
+    orgs
+    vdcs
+    vapps
+    vms
+    vapp_templates
+    images
+  end
+
+  private
+
+  def orgs
+    collector.orgs
+  end
+
+  def vdcs
+    collector.vdcs.each do |vdc|
+      persister.availability_zones.find_or_build(vdc.id).assign_attributes(
+        :ems_ref => vdc.id,
+        :name    => vdc.name
+      )
+    end
+  end
+
+  def vapps
+    collector.vapps.each do |vapp|
+      persister.orchestration_stacks.find_or_build(vapp.id).assign_attributes(
+        :ems_ref     => vapp.id,
+        :name        => vapp.name,
+        :description => vapp.name,
+        :status      => vapp.human_status,
+      )
+    end
+  end
+
+  def vms
+    collector.vms.each do |vm|
+      parsed_vm = persister.vms.find_or_build(vm[:vm].id).assign_attributes(
+        :uid_ems             => vm[:vm].id,
+        :ems_ref             => vm[:vm].id,
+        :name                => vm[:vm].name,
+        :hostname            => vm[:hostname],
+        :vendor              => 'vmware',
+        :raw_power_state     => vm[:vm].status,
+        :orchestration_stack => persister.orchestration_stacks.lazy_find(vm[:vm].vapp_id),
+      )
+
+      if (snapshot_resp = vm[:snapshot].fetch_path(:body, :Snapshot))
```

Comment:
Two things:

- we don't want to crash if `vm[:snapshot]` is `nil`
- I don't like `snapshot_resp` name, sounds so hacky

```ruby
if (resp = vm[:snapshot]) && (snapshot = resp.fetch_path(:body, :Snapshot))
  parsed_vm.snapshots = [
    persister.snapshots.find_or_build(parsed_vm.ems_ref).assign_attributes(
        ...
    )
  ]
end
```